src/main/java/com/tabnine/intellij/completions/LimitExceededLookupElement.java

```diff
@@ -0,0 +1,147 @@
+package com.tabnine.intellij.completions;
+
+import com.intellij.codeInsight.daemon.impl.HintRenderer;
+import com.intellij.codeInsight.lookup.LookupElement;
+import com.intellij.codeInsight.lookup.LookupEvent;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.diagnostic.ControlFlowException;
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.editor.Document;
+import com.intellij.openapi.editor.Editor;
+import com.intellij.openapi.editor.Inlay;
+import com.intellij.openapi.editor.colors.EditorColorsManager;
+import com.intellij.openapi.editor.event.DocumentEvent;
+import com.intellij.openapi.editor.event.DocumentListener;
+import com.intellij.openapi.editor.event.EditorMouseEvent;
+import com.intellij.openapi.editor.event.EditorMouseMotionListener;
+import com.intellij.openapi.ui.popup.Balloon;
+import com.intellij.openapi.ui.popup.JBPopupFactory;
+import com.intellij.openapi.util.Disposer;
+import com.intellij.ui.JBColor;
+import com.intellij.ui.awt.RelativePoint;
+import com.intellij.util.DocumentUtil;
+import com.tabnine.binary.BinaryRequestFacade;
+import com.tabnine.binary.requests.notifications.HoverBinaryRequest;
+import com.tabnine.binary.requests.notifications.HoverBinaryResponse;
+import com.tabnine.binary.requests.notifications.actions.HoverActionRequest;
+import com.tabnine.general.DependencyContainer;
+import com.tabnine.general.NotificationOption;
+import com.tabnine.general.StaticConfig;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import javax.swing.event.HyperlinkEvent;
+import javax.swing.event.HyperlinkListener;
+import java.awt.*;
+import java.awt.event.MouseEvent;
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class LimitExceededLookupElement extends InsertNothingLookupElement {
```

Comment:
I'm not comfortable approving the PR with this file. I do not understand all the concurrency flows here, and I feel that it could be refactored to multiple files that could be easier to understand.